lib/android/app/src/main/java/com/reactnativenavigation/anim/TopbarCollapsingBehavior.java

```diff
@@ -0,0 +1,81 @@
+package com.reactnativenavigation.anim;
+
+
+import android.view.View;
+
+import com.facebook.react.uimanager.events.EventDispatcher;
+import com.reactnativenavigation.interfaces.ScrollEventListener;
+import com.reactnativenavigation.utils.UiThread;
+import com.reactnativenavigation.views.TopBar;
+
+public class TopbarCollapsingBehavior {
+    private TopBar topBar;
+
+    private EventDispatcher eventDispatcher;
+    private ScrollEventListener scrollEventListener;
+    private boolean dragStarted;
+    private TopBarAnimator animator;
+
+    public TopbarCollapsingBehavior(EventDispatcher eventDispatcher, TopBar topBar) {
+        this.eventDispatcher = eventDispatcher;
+        this.topBar = topBar;
+        this.animator = new TopBarAnimator();
+    }
+
+    public void enableCollapsing() {
+        scrollEventListener = (new ScrollEventListener(new ScrollEventListener.OnVerticalScrollListener() {
+            @Override
+            public void onVerticalScroll(int scrollY, int oldScrollY) {
+                if (scrollY < 0) return;
+                if (!dragStarted) return;
+
+                int measuredHeight = topBar.getMeasuredHeight();
+                int diff = scrollY - oldScrollY;
+                if (Math.abs(diff) > measuredHeight) {
+                    diff = (Math.abs(diff) / diff) * measuredHeight;
+                }
+                float nextTranslation = topBar.getTranslationY() - diff;
+                if (diff < 0) {
```

Comment:
Whenever I have an `if else` block with more than 3 lines, I extract the logic into functions with explicit names. this makes the code much more readable and understandable.

```java
if (diff < 0) {
    animateDown();
} else {
    animateUp();
}
```